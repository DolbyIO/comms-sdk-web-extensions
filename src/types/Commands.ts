import { EventEmitter } from 'events';
import { Participant } from '@voxeet/voxeet-web-sdk/types/models/Participant';

export type ActionCommand = {
    action: string;
    payload: string;
};

export type TargetedCommand = {
    target: string[];
    message: string;
};

export type CommandsType = EventEmitter & {
    /**
     * Sends a command to a specific list of participants.
     * If the list is empty, the command will be received by all the participants.
     *
     * **Important:** the command will be received by all the participants at all time,
     * but will be filtered by this extension and the message will not be propagated
     * if the current participant is not in the target list.
     */
    send: (message: string, participants: Participant[]) => Promise<void>;
};
