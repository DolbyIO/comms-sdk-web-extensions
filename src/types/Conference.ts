import Conference from '@voxeet/voxeet-web-sdk/types/models/Conference';
import { JoinOptions, ListenOptions } from '@voxeet/voxeet-web-sdk/types/models/Options';
import { Participant } from '@voxeet/voxeet-web-sdk/types/models/Participant';

export type AttachMediaStreamToHTMLVideoElement = (participant: Participant) => HTMLVideoElement;

export type Conferencing = {
    /**
     * Gets an HTML video element for the specified participant.
     * @param participant Participant to display the stream from.
     * @returns an `HTMLVideoElement` element.
     */
    attachMediaStreamToHTMLVideoElement: AttachMediaStreamToHTMLVideoElement;

    /**
     * Rejoins the conference as a USER.
     * @param options - The additional options for the joining participant.
     * @returns the `Conference` object that represents the conference that was joined.
     */
    switchToUser: (options: JoinOptions) => Promise<Conference>;

    /**
     * Rejoins the conference as a LISTENER.
     * @param options - The additional options for the joining participant.
     * @returns the `Conference` object that represents the conference that was joined.
     */
    switchToListener: (options?: ListenOptions) => Promise<Conference>;
};

export type CreateConferencing = () => Conferencing;
