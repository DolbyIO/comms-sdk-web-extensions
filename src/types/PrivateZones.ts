import { SpatialPosition, SpatialScale, Vector } from '@voxeet/voxeet-web-sdk/types/models/SpatialAudio';
import { Participant } from '@voxeet/voxeet-web-sdk/types/models/Participant';

type SpatialDimension = Vector<number>;

/**
 * Model representing a zone.
 */
export type Zone = {
    /** Origin coordinates of the zone. */
    origin: SpatialPosition;
    /** The dimension of the zone. */
    dimension: SpatialDimension;
    /** The application's distance units or scale in application units per one meter. */
    scale: SpatialScale;
    /** Users' video is restricted in the boundaries of the zone. Default is false. */
    videoRestriction?: boolean;
};

export type Zones = Map<string, Zone>;

export type PrivateZones = {
    /**
     * Configures a spatial environment of an application, so the audio renderer understands which directions the application considers forward, up, and right and which units it uses for distance.
     *
     * This method is available only for participants who joined the conference with the [spatialAudio](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-joinoptions#spatialaudio) parameter enabled. Otherwise, SDK triggers [UnsupportedError](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-unsupportederror).
     *
     * If not called, the SDK uses the default spatial environment, which consists of the following values:
     *
     * - `forward` = (0, 0, 1), where +Z axis is in front
     * - `up` = (0, 1, 0), where +Y axis is above
     * - `right` = (1, 0, 0), where +X axis is to the right
     * - `scale` = (1, 1, 1), where one unit on any axis is 1 meter
     *
     * If sending the updated positions to the server fails, the SDK generates the ConferenceService event error that includes [SpatialAudioError](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-spatialaudioerror).
     *
     * @param scale - The application's distance units or scale in application units per one meter. The value must be greater than 0. Otherwise, SDK emits [ParameterError](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-parametererror).
     * @param forward - A vector describing the direction the application considers as forward. The value must be orthogonal to up and right. Otherwise, SDK emits [ParameterError](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-parametererror).
     * @param up - A vector describing the direction the application considers as up. Must be orthogonal to forward and right. Otherwise, SDK emits [ParameterError](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-parametererror).
     * @param right - A vector describing the direction the application considers as right. Must be orthogonal to forward and up. Otherwise, SDK emits [ParameterError](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-parametererror).
     */
    setSpatialEnvironment: (scale: SpatialScale, forward: SpatialPosition, up: SpatialPosition, right: SpatialPosition) => void;

    /**
     * Creates a private audio zone.
     * @param zone private zone to create.
     * @returns the unique identifier for this newly created private zone.
     */
    createZone: (zone: Zone) => Promise<string>;

    /**
     * Gets the list of private zones.
     * @returns a dictionary of private zones.
     */
    zones: Zones;

    /**
     * Deletes a private zone.
     * @param id identifier of the private zone to delete.
     */
    deleteZone: (zoneId: string) => Promise<void>;

    /**
     * Updates a private zone.
     * @param id identifier of the private zone to update.
     * @param zone new settings for the private zone.
     */
    updateZone: (zoneId: string, zone: Zone) => Promise<void>;

    /**
     * Sets a participant's position in space to enable the spatial audio experience during a Dolby Voice conference. This method is available only for participants who joined the conference with the [spatialAudio](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-joinoptions#spatialaudio) parameter enabled. Otherwise, SDK triggers [UnsupportedError](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-unsupportederror). Depending on the specified participant in the `participant` parameter, the setSpatialPosition method impacts the location from which audio is heard or from which audio is rendered:
     *
     * - When the specified participant is the local participant, setSpatialPosition sets a location from which the local participant listens to a conference. If the local participant does not have an established location, the participant hears audio from the default location (0, 0, 0).
     *
     * - When the specified participant is a remote participant, setSpatialPosition ensures the remote participant's audio is rendered from the specified position in space. If the remote participant does not have an established location, the participant does not have a default position and will remain muted until a position is specified.
     *
     * For example, if a local participant Eric, who does not have a set direction, calls setSpatialPosition(VoxeetSDK.session.participant, {x:3,y:0,z:0}), Eric hears audio from the position (3,0,0). If Eric also calls setSpatialPosition(Sophia, {x:7,y:1,z:2}), he hears Sophia from the position (7,1,2). In this case, Eric hears Sophia 4 meters to the right, 1 meter above, and 2 meters in front.
     *
     * If sending the updated positions to the server fails, the SDK generates the ConferenceService event error that includes [SpatialAudioError](https://docs.dolby.io/communications-apis/docs/js-client-sdk-model-spatialaudioerror).
     *
     * @param participant - The selected participant. Using the local participant sets the location from which the participant will hear a conference. Using a remote participant sets the position from which the participant's audio will be rendered.
     * @param position - The participants' audio location.
     */
    setSpatialPosition: (participant: Participant, position: NonNullable<SpatialPosition>) => Promise<void>;
};

export type CreatePrivateZones = () => PrivateZones;
