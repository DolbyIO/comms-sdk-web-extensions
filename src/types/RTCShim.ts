export type RTCShim = {
    attachMediaStream: (elem: HTMLVideoElement, stream: MediaStream) => any;
};
