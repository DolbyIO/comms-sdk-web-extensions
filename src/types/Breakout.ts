import { Participant } from '@voxeet/voxeet-web-sdk/types/models/Participant';

import { CommandsType } from './Commands';

/**
 * Model representing a breakout room.
 */
export type BreakoutRoom = {
    name: string;
    participants: Participant[];
};

export type BreakoutRoomCommandPayload = {
    name: string;
    participants: string[];
};

export type BreakoutRooms = Map<string, BreakoutRoom>;

export type Breakout = {
    /**
     * Gets the list of breakout rooms.
     * @returns a dictionary of breakout rooms.
     */
    breakoutRooms: BreakoutRooms;

    /**
     * Creates a new breakout room.
     * @param room - Breakout room to create.
     * @returns the unique identifier for this newly created breakout room.
     */
    createRoom: (room: BreakoutRoom) => Promise<string>;

    /**
     * Closes a breakout room and move all participants to the main room.
     * @param id - Identifier of the breakout room to close.
     */
    closeRoom: (id: string) => Promise<void>;

    /**
     * Moves a list of participants into a breakout room.
     * @param roomId - Identifier of the breakout room to move the participants to.
     *                 If this value is `null` then the participants will be moved to the main room.
     * @param participants - List of participants to move.
     */
    moveTo: (roomId: string | null, participants: Participant[]) => Promise<void>;
};

export type CreateBreakout = (commands: CommandsType) => Breakout;
