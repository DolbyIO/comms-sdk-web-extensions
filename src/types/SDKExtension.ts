import { SpatialAudio } from './SpatialAudio';
import { Conferencing } from './Conference';

export type Extension = {
    /**
     * Retrieves the Conference instance that allows to access helpers for the conference.
     */
    conference: Conferencing;

    /**
     * Retrieves the SpatialAudio instance that allows to manipulate the spatial audio scene.
     */
    spatialAudio: SpatialAudio;
};

export type CreateExtension = () => Extension;
