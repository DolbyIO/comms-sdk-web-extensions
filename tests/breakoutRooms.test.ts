import { EventEmitter } from 'events';
import './__mocks__/MediaStream.mock';
import './__mocks__/navigator.mock';
import createBreakout from '../src/breakout';
import Commands from '../src/commands';
import { Breakout, BreakoutRoom } from '../src/types/Breakout';

var mockedParticipant: any = { id: '000' };
const mockedCommandSend = jest.fn((message: string) => new Promise((resolve) => resolve({})));

const conferenceMock = new EventEmitter();
var ps: Map<string, any> = new Map<string, any>();
ps.set('000', { id: '000' });
ps.set('111', { id: '111' });
ps.set('222', { id: '222' });

(<any>conferenceMock).participants = ps;

const commandMock = new EventEmitter();
(<any>commandMock).send = mockedCommandSend;

jest.mock('@voxeet/voxeet-web-sdk', () => ({
    get conference() {
        return conferenceMock;
    },
    get command() {
        return commandMock;
    },
    get notification() {
        return new EventEmitter();
    },
    get session() {
        return {
            get participant() {
                return mockedParticipant;
            },
        };
    },
}));

describe('breakoutRooms test suite', () => {
    let breakout: Breakout;

    beforeEach(() => {
        const commands = new Commands();
        breakout = createBreakout(commands);
    });

    test('breakoutRooms should match the snapshot', () => {
        expect(breakout).toMatchSnapshot();
    });

    test('breakoutRooms createRoom', async () => {
        const br: BreakoutRoom = {
            name: 'first',
            participants: [],
        };
        const roomId = await breakout.createRoom(br);

        expect(breakout.breakoutRooms.size).toEqual(1);

        await breakout.closeRoom(roomId);

        expect(breakout.breakoutRooms.size).toEqual(0);
    });

    test('breakoutRooms moveTo', async () => {
        const br: BreakoutRoom = {
            name: 'first',
            participants: [],
        };
        const roomId = await breakout.createRoom(br);

        // Create another room
        const br2: BreakoutRoom = {
            name: 'second',
            participants: [],
        };
        await breakout.createRoom(br2);

        await breakout.moveTo(roomId, [<any>{ id: '000' }]);

        expect(breakout.breakoutRooms.size).toEqual(2);
        expect(breakout.breakoutRooms.get(roomId)?.participants.length).toEqual(1);
    });

    test('Exceptions - moveTo', async () => {
        const wrongId = 'wrong-id';

        try {
            await breakout.moveTo(wrongId, []);
        } catch (e) {
            expect(e).toEqual(`The breakout room with the id ${wrongId} does not exist.`);
        }
    });

    test('breakoutRooms moveToMainRoom', async () => {
        const br: BreakoutRoom = {
            name: 'first',
            participants: [<any>{ id: '000' }],
        };
        const roomId = await breakout.createRoom(br);

        expect(breakout.breakoutRooms.size).toEqual(1);
        expect(breakout.breakoutRooms.get(roomId)?.participants.length).toEqual(1);

        // Move to the main room
        await breakout.moveToMainRoom([<any>{ id: '000' }]);

        expect(breakout.breakoutRooms.size).toEqual(1);
        expect(breakout.breakoutRooms.get(roomId)?.participants.length).toEqual(0);
    });

    test('Exceptions - closeRoom', async () => {
        const wrongId = 'wrong-id';

        try {
            await breakout.closeRoom(wrongId);
        } catch (e) {
            expect(e).toEqual(`The breakout room with the id ${wrongId} does not exist.`);
        }
    });
});
