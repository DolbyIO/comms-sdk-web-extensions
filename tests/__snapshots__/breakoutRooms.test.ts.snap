// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breakoutRooms test suite breakoutRooms should match the snapshot 1`] = `
{
  "breakoutRooms": Map {},
  "closeRoom": [Function],
  "createRoom": [Function],
  "initialize": [Function],
  "moveTo": [Function],
  "moveToMainRoom": [Function],
}
`;
